syntax = "proto3";
option cc_enable_arenas = true;

package Ydb.Formats;
option java_package = "com.yandex.ydb.formats";

message ArrowBatchSettings {
    bytes schema = 1;
}

message CsvSettings {
    message Quoting {
        bool disabled = 1;
        bytes quote_char = 2;
        bool double_quote_disabled = 3;
    }

    // Number of rows to skip before CSV data. It should be present only in the first upsert of CSV file.
    uint32 skip_rows = 1;
    // Fields delimiter in CSV file. It's "," if not set.
    bytes delimiter = 2;
    // String value that would be interpreted as NULL.
    bytes null_value = 3;
    // First not skipped line is a CSV header (list of column names).
    bool header = 4;
    Quoting quoting = 5;
}

enum SchemaInclusionMode {
    SCHEMA_INCLUSION_MODE_UNSPECIFIED = 0;
    SCHEMA_INCLUSION_MODE_ALWAYS = 10;        // Always include schema in the response stream.
    SCHEMA_INCLUSION_MODE_FIRST_ONLY = 20;    // Include schema only in the first part of the response stream.
}

/**
 * Ydb.Value output format for Query.ExecuteQueryRequest
*/
message ValueOutputFormat {
    // If Ydb.ResultSet schema is not needed, it can be omitted in the response stream.
    SchemaInclusionMode schema_inclusion_mode = 1;
}

message ValueOutputFormatMeta {}

/**
 * Arrow output format for Query.ExecuteQueryRequest
*/
message ArrowOutputFormat {
    message Compression {
        enum Type {
            COMPRESSION_UNSPECIFIED = 0;
            COMPRESSION_NONE = 1;
            COMPRESSION_ZSTD = 2;
            COMPRESSION_LZ4_FRAME = 3;
        }

        Type type = 1;
        int32 level = 2;
    }

    // If Ydb.ResultSet schema is not needed, it can be omitted in the response stream.
    SchemaInclusionMode schema_inclusion_mode = 1;
    // Compression codec to use.
    Compression compression = 2;
}

message ArrowOutputFormatMeta {
    // It's filled only for SchemaInclusionMode ALWAYS or FIRST_ONLY (first part of the response stream).
    bytes schema = 1;
}
